File: 66. Plus One/PlusOne.py
from typing import List


class Solution:
    # O(n) time,
    # O(1) space,
    # Approach: array, 
    def plusOne(self, digits: List[int]) -> List[int]:
        
        left_over = True
        
        for i in range(len(digits)-1, -1, -1):
            if not left_over:
                break
            
            left_over = (digits[i] == 9)
            digits[i] = 0 if digits[i] == 9 else digits[i] + 1
        
        return ([1] + digits) if left_over else digits

File: 66. Plus One/plusOne.go
package PlusOne


// O(n) time,
// O(n) space,
// Approach: simualtion, math, array
func plusOne(digits []int) []int {
    right := len(digits)-1
    
    leftOver := true
    for right >= 0 && leftOver {
        digits[right] += 1
        if digits[right] > 9 {
            leftOver = true
        } else {
            leftOver = false
        }
        right--
    }
    
    if leftOver {
        digits = append([]int{1}, digits...)
    }
    
    return digits
}